src/scheduler/data_types.h

```diff
@@ -165,6 +174,83 @@ typedef sch_resource_t usage_t;
 typedef void event_ptr_t;
 typedef int (*event_func_t)(event_ptr_t*, void *);
 
+struct th_task_info
+{
+	int task_id;							/* task id, should be set by main thread */
+	enum thread_task_type task_type;		/* task type */
+	void *thread_data;					/* data for the worker thread to execute the task */
+};
+
+struct th_data_nd_eligible
+{
+	resource_resv *resresv;
+	place *pl;
+	schd_error *err;
+	char *exclerr_buf;
+	node_info **ninfo_arr;
+	int sidx;
+	int eidx;
+};
+
+struct th_data_dup_nd_info
+{
+	unsigned int err:1;
```

Comment:
It should be pretty easy to change if we have a need for this in the future, how about we change it if/when it's needed?

Reply:
why not now? all we have to do is that we make it int and assign it a generic error number (which is most cases is going to be a malloc error)

Reply:
thinking about this more, i think if there's a need to convey more information than just "error happened", then we should be passing a schd_error object, which I actually do use for check_node_array_eligibility, its thread data type 'th_data_nd_eligible' has a schd_error object.

Reply:
Great! in that case, can you make it consistent across other datastructures?

Reply:
I think I'm not explaining this well. Each 'th_data_' object is just a way to pass data to a work task. Since each work task is unique, each data object is also unique, this is very much like different functions requiring specific arguments. So, what you are asking me to do is similar to saying that pass an argument to a function which it won't use today but might use in the future. We don't pass a schd_error argument to all scheduler functions, right? Similarly, not all functions return multiple return values right? Most of them just return a 0 or a 1, which is exactly what I want to achieve with the 'error' bit. Do you see what I'm saying now?

Reply:
Ok I understand your point. What I do not understand is that what benefit do we get if it is kept as a bit field in some structures.

Reply:
As I said, there was a need to convey whether or not an error happened for some tasks, just like some functions return 0 or 1. A bit field serves that purpose, that's why I used it. Why spend more memory if we don't need to?

Reply:
I don’t think you are saving any memory by making it a bit field. Your structure will be padded. Any other reason?

Reply:
No other reason. There are many other scheduler structs which use this syntax for boolean flags, can you please clarify why you are being so particular about not using it here?

Reply:
You use bit fields when you want to more than one thing at a time. Many other data structures in scheduler use it to convey flags (like user limit, group limit etc.) they don’t use it to convey errors. If errors were kept as flags we will be modifying structure every time there is a new type of error added. It just does not make any sense.

Reply:
but it's not really supposed to convey the type of error that happened, it's just like a return value. I'll do 2 things, rename it to "retval" and convert it to an int, does that sound good? Please try to look at this struct as that function's arguments and return type, it will not be used anywhere else except that one function.